helpers/embeds.js

```diff
@@ -0,0 +1,82 @@
+import { html, render } from 'lit';
+import { findComposedAncestor } from '../helpers/dom.js';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+const embedTypeAttributeName = 'data-d2l-embed-type';
+
+class RenderNode {
+	#childNodes;
+	#node;
+
+	constructor(node) {
+		this.#node = node;
+		this.#childNodes = [];
+	}
+
+	addChildNode(childNode) {
+		this.#childNodes.push(childNode);
+	}
+
+	async renderNode(options) {
+		await Promise.all(this.#childNodes.map(childNode => childNode.renderNode(options)));
+
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', this.#node.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const fragment = new DocumentFragment();
+		this.#node.childNodes.forEach(node => {
+			const temp = node.cloneNode(true);
+			fragment.appendChild(temp);
+		});
+
+		let props = this.#node.dataset.d2lEmbedData;
+		try {
+			props = JSON.parse(this.#node.dataset.d2lEmbedData);
+		} catch (e) { /* empty */ }
+
+		const embed = await embedRendererPlugin.renderView(fragment, props || {}, options || {});
+		if (embed) {
+			render(embed, this.#node.parentNode);
+			this.#node.remove();
+		}
+	}
+
+}
+
+export async function createEmbedPlaceholder(embedType, options) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(options);
+	/* NOT IMPLEMENTED */
+}
+
+export async function renderEmbeds(node, options) {
+	const elems = node.querySelectorAll(`div[${embedTypeAttributeName}], span[${embedTypeAttributeName}]`);
+	if (elems.length === 0) return;
+
+	const renderNodes = new Map();
+	const rootNodes = [];
+	const constructRenderNode = elem => {
+		if (renderNodes.has(elem)) return;
```

Comment:
Is there a scenario where this would ever be `true`? `renderNodes` is initialized in this method, so it seems like even if this were to run a second time, it would be impossible case. Is this just being extra careful?

Reply:
It's just being extra careful, yeah. This shouldn't be possible for the same reason this algorithm (hopefully) works in the first place; since we're iterating over a static NodeList that's guaranteed to be in DOM order, we can never get a duplicate.